hazelcast/src/main/java/com/hazelcast/partition/strategy/AttributePartitioningStrategy.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) 2008-2023, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.partition.strategy;
+
+import com.hazelcast.core.HazelcastJsonValue;
+import com.hazelcast.internal.serialization.SerializableByConvention;
+import com.hazelcast.internal.serialization.impl.GenericRecordQueryReader;
+import com.hazelcast.internal.serialization.impl.InternalGenericRecord;
+import com.hazelcast.jet.impl.util.ReflectionUtils;
+import com.hazelcast.nio.serialization.DataSerializable;
+import com.hazelcast.partition.PartitioningStrategy;
+import com.hazelcast.query.impl.getters.JsonGetter;
+
+import java.io.IOException;
+import java.io.Serializable;
+
+@SerializableByConvention
+public class AttributePartitioningStrategy implements PartitioningStrategy {
+
+    private String[] attributes;
+
+    public AttributePartitioningStrategy(String... attributes) {
+        this.attributes = attributes;
+    }
+
+    @Override
+    public Object getPartitionKey(final Object key) {
+        final Object[] values = new Object[attributes.length];
```

Comment:
I do not see much value in the preallocation of the array, and the control flow is convoluted. It would be much simpler if `extractFromXXX` methods returned `values` or null.

Reply:
much simpler now :)